exception thrown around internal compilation unit cache

No idea how and when this happened, but I saw this in the log output, so let's capture this here and take a look.

```
09:55:46.069 ERROR [ForkJoinPool.commonPool-worker-1] o.s.i.v.b.j.utils.CompilationUnitCache   : 
java.util.concurrent.CompletionException: java.util.concurrent.CancellationException
	at java.base/java.util.concurrent.CompletableFuture.wrapInCompletionException(CompletableFuture.java:323)
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:376)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:391)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:729)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:531)
	at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2554)
	at org.springframework.ide.vscode.boot.java.utils.CompilationUnitCache$1.onRemoval(CompilationUnitCache.java:106)
	at com.google.common.cache.LocalCache.processPendingNotifications(LocalCache.java:1814)
	at com.google.common.cache.LocalCache$Segment.runUnlockedCleanup(LocalCache.java:3467)
	at com.google.common.cache.LocalCache$Segment.postWriteCleanup(LocalCache.java:3443)
	at com.google.common.cache.LocalCache$Segment.remove(LocalCache.java:3077)
	at com.google.common.cache.LocalCache.remove(LocalCache.java:4251)
	at com.google.common.cache.LocalCache.invalidateAll(LocalCache.java:4292)
	at com.google.common.cache.LocalCache$LocalManualCache.invalidateAll(LocalCache.java:4896)
	at org.springframework.ide.vscode.boot.java.utils.CompilationUnitCache.invalidateProject(CompilationUnitCache.java:389)
	at org.springframework.ide.vscode.boot.java.utils.CompilationUnitCache.lambda$3(CompilationUnitCache.java:171)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1825)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1817)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:511)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1450)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2019)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
Caused by: java.util.concurrent.CancellationException
	at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2552)
	... 16 common frames omitted
```

--- a/headless-services/spring-boot-language-server/src/main/java/org/springframework/ide/vscode/boot/java/utils/CompilationUnitCache.java
+++ b/headless-services/spring-boot-language-server/src/main/java/org/springframework/ide/vscode/boot/java/utils/CompilationUnitCache.java
@@ -294,7 +294,7 @@ public final class CompilationUnitCache implements DocumentContentProvider {
 					}
 				})
 				.exceptionally(t -> {
-					if (!(t instanceof CancellationException)) {
+					if (!isCancellation(t)) {
 						logger.error("", t);
 					}
 					synchronized(CompilationUnitCache.this) {
@@ -304,6 +304,13 @@ public final class CompilationUnitCache implements DocumentContentProvider {
 				});
 		}
 		return cuFuture;
+	}
+
+	static boolean isCancellation(Throwable t) {
+		while (t instanceof CompletionException && t.getCause() != null) {
+			t = t.getCause();
+		}
+		return t instanceof CancellationException;
 	}
 
 	public static CompilationUnit parse2(char[] source, String docURI, String unitName, IJavaProject project) throws Exception {

--- a/headless-services/spring-boot-language-server/src/test/java/org/springframework/ide/vscode/boot/java/utils/CompilationUnitCacheExceptionTest.java
+++ b/headless-services/spring-boot-language-server/src/test/java/org/springframework/ide/vscode/boot/java/utils/CompilationUnitCacheExceptionTest.java
@@ -1,0 +1,38 @@
+/*******************************************************************************
+ * Copyright (c) 2026 Pivotal, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Pivotal, Inc. - initial API and implementation
+ *******************************************************************************/
+package org.springframework.ide.vscode.boot.java.utils;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletionException;
+
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests for exception classification in the compilation unit cache.
+ *
+ * @author Pavel Zaitsev
+ */
+class CompilationUnitCacheExceptionTest {
+
+	@Test
+	void recognizesCancellationWrappedByCompletionStage() {
+		assertTrue(CompilationUnitCache.isCancellation(new CompletionException(new CancellationException())));
+	}
+
+	@Test
+	void doesNotTreatRealExceptionalCompletionAsCancellation() {
+		assertFalse(CompilationUnitCache.isCancellation(new CompletionException(new IllegalStateException())));
+	}
+
+}

